sdjournal/read.go

```diff
@@ -101,35 +103,53 @@ func NewJournalReader(config JournalReaderConfig) (*JournalReader, error) {
 	return r, nil
 }
 
+// Read reads entries from the journal. Read follows the Reader interface so
+// it must be able to read a specific amount of bytes. Journald on the other
+// hand only allows us to read full entries of arbitrary size (without byte
+// granularity). JournalReader is therefore internally buffering entries that
+// don't fit in the read buffer. Callers should keep calling until 0 and/or an
+// error is returned.
 func (r *JournalReader) Read(b []byte) (int, error) {
 	var err error
-	var c int
 
-	// Advance the journal cursor
-	c, err = r.journal.Next()
+	if r.msgReader == nil {
+		var c int
 
-	// An unexpected error
-	if err != nil {
-		return 0, err
-	}
+		// Advance the journal cursor. It has to be called at least one time
+		// before reading
+		c, err = r.journal.Next()
 
-	// EOF detection
-	if c == 0 {
-		return 0, io.EOF
-	}
+		// An unexpected error
+		if err != nil {
+			return 0, err
+		}
 
-	// Build a message
-	var msg string
-	msg, err = r.buildMessage()
+		// EOF detection
+		if c == 0 {
+			return 0, io.EOF
+		}
 
-	if err != nil {
-		return 0, err
+		// Build a message
+		var msg string
+		msg, err = r.buildMessage()
+
+		if err != nil {
+			return 0, err
+		}
+		r.msgReader = strings.NewReader(msg)
 	}
 
 	// Copy and return the message
-	copy(b, []byte(msg))
+	var sz int
+	sz, err = r.msgReader.Read(b)
+	if err != nil {
+		return sz, err
```

Comment:
This may return (sz > 0, EOF) once the buffer is completely consumed, at least in theory according to Reader interface.